### DISCUSSION TITLE
Batch upsert efficiency

### DISCUSSION THREAD

**Discussion by alevinetx** [simolus3/drift#2035] [Q&A] [answered]
I have a list of `ModelCompanion`s.  Some items are new (they haven't been saved and have no id), some exist and have changed values, and some may exist and not have any changed values.

As I understand it, there's a cost for each ffi interaction, so batches are useful.  

When it comes to doing a batch with the above dataset,  which would be the most efficient?

 Thank you!

where `next` is `List<ModelCompanion>`

1:
```
  await db.batch((batch) {

    for (var f in next)
      {
        if (f.id.present)
          {
            batch.replace(table, f);
          }
        else
          {
              batch.insertAllOnConflictUpdate(table, [f]);
          }
      } //for
  }  // batch
  );  //await
```

2:
```
  await db.batch((batch) {
     batch.insertAllOnConflictUpdate(table, next);
  }  // batch
  );  //await
```

3:
  a continuation of 1, but creating two lists first and then passing a full list to update/insert?

4:
  an entirely different manner

**Comment by simolus3** [comment, marked as answer]
> As I understand it, there's a cost for each ffi interaction, so batches are useful.

Compared to using `drift_sqflite` with method channels, the actual overhead of `dart:ffi` is pretty small. Using a batch mainly improves performance for two reasons:

1. It uses a transaction, which means that sqlite3 doesn't have to do a write + flush for every individual statement.
2. It re-uses prepared statements, which means that you avoid duplicate parsing costs when running similar SQL queries that only differ in their variables.

In your case, I think the performance differences of the three options should be marginal. It definitely doesn't have much of an impact in drift. It could be that the `UPDATE` statement generated by `replace` is slightly more efficient than `INSERT ON CONFLICT` inside sqlite3, but that typically doesn't make up much of the execution time.

You could just try it and see if it makes a difference (I don't think it does - if there turns out to be a performance difference I'd love to take a look at it). But otherwise I'd just optimize this for simplicity and go with option 2.

**Reply by alevinetx** [reply]
Thank you for the insight.

While my datasets aren't large enough yet to notice one being preferential over another, the nerd likes to future proof.
